text/0000-lifecycle-pkg-refactor

```diff
@@ -0,0 +1,174 @@
+
+With the current package structure, it can be challenging to determine where to put new code when new features are added. Various patterns for code organization have been tried. If we can identify which patterns have been most useful, and perpetuate those going forward, the lifecycle over time will become more organized and pleasant to work with.
+
+# What it is
+[what-it-is]: #what-it-is
+
+The target personas for this RFC are lifecycle contributors and lifecycle library consumers.
+
+Here is the current package structure:
+
+```
+├── api                         # helpers for working with api versions - mostly absent CNB logic except for Buildpack and Platform vars
+├── archive                     # helpers for working with tars - no CNB logic; mostly used by layers package
+├── auth                        # a wrapper around GGCR’s authn.Keychain that also reads the CNB_REGISTRY_AUTH env var; used by cmd package
+├── buildpack                   # logic pertaining to the Buildpack API
+├── cache                       # 3 implementations of the Cache interface; used by cmd package
+├── cmd
+│   ├── launcher
+│   ├── lifecycle
+├── env                         # helpers for constructing the environment for a buildpack during build, or for an app during launch
+├── image                       # helpers for working with registries and images
+├── internal
+│   ├── encoding                # WriteTOML and WriteJSON
+│   ├── io                      # Copy
+│   ├── layer                   # metadata restorer, sbom restorer
+│   └── str                     # Compare
+├── launch                      # logic for launching a process; used by cmd/launcher
+├── layers                      # logic for reading and writing layers; mostly used by exporter - also elsewhere for the concept of layers.Slice
+├── platform                    # contains structs representing platform spec’d data formats and exit code logic
+├── priv                        # logic for running the lifecycle as a particular user; used by cmd package
+├── analyzer.go                 # lifecycle logic lives in the top level
+├── detector.go
+├── rebaser.go
+├── ...
+```
+
+Here is the proposed package structure:
+
+```
+├── cmd
+│    ├── launcher
+│    ├── lifecycle
+├── internal
+│    ├── buildpack              # services for handling logic that branches on buildpack api - e.g., bom validator
+│    ├── encoding
+│    ├── image
+│    ├── io
+│    ├── platform               # services for handling logic that branches on platform api - e.g., metadata restorer, sbom restorer, potentially a new sbom copier, etc.
+│    └── str
+├── pkg
+│    ├── api                    # old api package with CNB logic that indicates supported and deprecated apis moved elsewhere
+│    ├── archive
+│    ├── buildpack
+│    │   ├── env
+│    │   ├── launch
```

Comment:
Actually nothing in `buildpack` imports this code. Maybe it should live at `pkg/launch`.